### PR TITLE
Fix/not update menu with i18n (#22594)

### PR DESCRIPTION
**Pull request by juggernautjp**
### Identify the Bug

Issue [#22594](https://github.com/atom/atom/issues/22594)

### Description of the Change

In regard to MenuManager.add(items) in atom/src/menu-manager.coffee.

Atom Core and [`atom-i18n`](https://github.com/liuderchi/atom-i18n) package use the same `label` member of the argument items, 
which is mentioned above.
After `atom-i18n` localize menus and submenus, Atom Core can not update them correctly.

So, I added the class member `id` to the argument `items`, which means a member of class MenuItem in Electron.
After `atom-i18n` localize `label`, Atom Core use `id` (instead of `label`, which is current implementation) to update menus and submenus.

#### Modified files

1. atom/src/menu-helpers.js
2. atom/src/menu-manager.coffee
3. atom/src/reopen-project-menu-manager.js
4. atom/src/main-process/application-menu.js

### Alternate Designs

I used `sublabel` instead of `id`. But strings of `sublabel` appear under the strings of `label`.

### Possible Drawbacks

1. The menus, which are created from menus/*.cson, of Atom Core and other packages are not updated correctly.
2. I have not yet create specs files, because I need to study how to write specs.
3. I tested on only Windows. This fix should be verified on Linux and macOS.

### Verification Process

I imported [`atom-i18n`](https://github.com/liuderchi/atom-i18n) package, and set locale to Japanese.
1. The menubar has "ファイル(F)" and no "File" menu, and "Reopen Project" menu has correct projects.
2. The "ヘルプ(H)" menu has "アップデートを確認", which means "Check for Update".

### Release Notes

When Atom is localized with the atom-i18n package, the "File" menu is no longer added to the menu bar and "Check for Update" is localized and displayed in the "Help" menu on Windows.

### CODE REVIEW

**Comment by icecream17** [comment]
Problems with adding id on top of label:
id is redundant, ideally just use the label

Problems with replacing label with id:
If you rename `label` to `id` the code would still do the same thing

Sure, it does extend `id` from `MenuItem`, and `id` can be used to identify, but `label` can also be used to identify, and the sort in `src/menu-manager.coffee` takes any property name. So I think there's a deeper different reason why the bug is happening.

Bug:
> After atom-i18n localize menus and submenus, Atom Core can not update them correctly.

Maybe the problem is that the menus and submenus are localized? But it's also bad that atom core can't update correctly.

Also the tests aren't updated, so they're failing:

```shell
  Example fail
Expected { label : 'A', submenu : [ { label : 'B', command : 'b', id : 'B' } ], id : 'A' } to equal { label : 'A', submenu : [ { label : 'B', command : 'b' } ] }.
```

**Comment by juggernautjp** [comment]
## 1. Atom Internationalization (i18n)

For globally used applications, such as VS Code and Atom editor, it is very important that they are localized in their own languages. In Japan, VS Code is almost all localized, so it has become more popular than Atom editor, last year.
The Atom Editor can be localized with the `atom-i18n` package, but it cannot localize menus or some strings that appear in menus. That's because those strings are hard-coded into the Atom Core's source code.

For internationalization of Atom, the `label` item of Menu should not be used in the merge/unmerge, etc.
Atom Core programmers should use `id` instead of `label` for Menu manipulation, in order for `atom-i18n` package to be able to localize their strings.


## 2. Fix the bug of CI error

I fixed the bug of the following CI bugs, and pass the CI test.

1. [error] Tests for core-render-process in G:/Downloads/atom/spec/menu-manager-spec.js. failed.
2. [error] Tests for core-render-process in G:/Downloads/atom/spec/context-menu-manager-spec.js. failed.

File: src/menu-manager.coffee [hunk 1]
```diff
@@ -78,17 +78,20 @@ class MenuManager
...
   # ```
   #
   # * `items` An {Array} of menu item {Object}s containing the keys:
-  #   * `label` The {String} menu label.
+  #   * `label` The {String} menu label, which is used for localiztion (in `atom-i18n` package).
   #   * `submenu` An optional {Array} of sub menu items.
   #   * `command` An optional {String} command to trigger when the item is
   #     clicked.
   #
+  #   * `id` (internal) The {String} menu id, which is used only in javascript code, is not used in menu template.
```
Comment (sadick254) [inline src/menu-manager.coffee:92]:
```suggestion
  #   * `id` (internal) The {String} menu id, not used in menu template.
```

File: src/menu-helpers.js [hunk 2]
```diff
@@ -53,14 +54,16 @@ function unmerge(menu, item) {
   }
 }
 
-function findMatchingItemIndex(menu, { type, label, submenu }) {
+// function findMatchingItemIndex(menu, { type, label, submenu }) {
```
Comment (sadick254) [inline src/menu-helpers.js:57]:
```suggestion
```

File: src/menu-helpers.js [hunk 3]
```diff
@@ -53,14 +54,16 @@ function unmerge(menu, item) {
...
+function findMatchingItemIndex(menu, { type, id, submenu }) {
   if (type === 'separator') {
     return -1;
   }
   for (let index = 0; index < menu.length; index++) {
     const item = menu[index];
     if (
-      normalizeLabel(item.label) === normalizeLabel(label) &&
+      // normalizeLabel(item.label) === normalizeLabel(label) &&
```
Comment (sadick254) [inline src/menu-helpers.js:65]:
```suggestion
```

File: src/menu-helpers.js [hunk 4]
```diff
@@ -93,6 +97,11 @@ function cloneMenuItem(item) {
     'beforeGroupContaining',
     'afterGroupContaining'
   );
+  if (item.id === null || item.id === undefined) {
+    // item.id is internally used by the function defined in this file, because `atom-i18n` localize item.label.
+    // cloneMenuItem() create menus from template (menu_{darwin, linux, win32}.cson), then `atom-i18n` can localize menu.
```
Comment (sadick254) [inline src/menu-helpers.js:102]:
```suggestion
```

**Review by sadick254 — CHANGES_REQUESTED** [review]

File: src/menu-manager.coffee [hunk 5]
```diff
@@ -201,7 +204,8 @@ class MenuManager
       []
 
   sortPackagesMenu: ->
-    packagesMenu = _.find @template, ({label}) -> MenuHelpers.normalizeLabel(label) is 'Packages'
+    # packagesMenu = _.find @template, ({label}) -> MenuHelpers.normalizeLabel(label) is 'Packages'
```
[the remark this answers is not in the document]
Comment (juggernautjp) [reply src/menu-manager.coffee:207]: OK. I delete.

File: src/menu-manager.coffee [hunk 6]
```diff
@@ -78,17 +78,20 @@ class MenuManager
...
   #       label: 'Hello'
-  #       submenu : [{label: 'World!', command: 'hello:world'}]
+  #       submenu : [{label: 'World!', id: 'World!', command: 'hello:world'}]
   #     }
   #   ]
   # ```
   #
   # * `items` An {Array} of menu item {Object}s containing the keys:
-  #   * `label` The {String} menu label.
+  #   * `label` The {String} menu label, which is used for localiztion (in `atom-i18n` package).
```
[the remark this answers is not in the document]
Comment (juggernautjp) [reply src/menu-manager.coffee:87]: I agree. I delete it.

File: src/menu-manager.coffee (hunk 1 above)
Comment (juggernautjp) [reply src/menu-manager.coffee:92]: I see. I delete.

**Review by juggernautjp — COMMENTED** [review]
I agree.

File: src/menu-helpers.js (hunk 2 above)
Comment (juggernautjp) [reply src/menu-helpers.js]: I see. I delete the line, according to your advice

File: src/menu-helpers.js (hunk 3 above)
Comment (juggernautjp) [reply src/menu-helpers.js]: I see. I delete the line, according to your advice.

File: src/menu-helpers.js (hunk 4 above)
Comment (juggernautjp) [reply src/menu-helpers.js]: I see. I delete the lines, according to your advice

File: src/main-process/application-menu.js [hunk 7]
```diff
@@ -115,16 +115,20 @@ module.exports = class ApplicationMenu {
   showUpdateMenuItem(state) {
     const items = this.flattenMenuItems(this.menu);
     const checkForUpdateItem = items.find(
-      ({ label }) => label === 'Check for Update'
+      // ({ label }) => label === 'Check for Update'
```
Comment (sadick254) [inline src/main-process/application-menu.js:118]:
```suggestion
```

File: src/main-process/application-menu.js [hunk 8]
```diff
@@ -115,16 +115,20 @@ module.exports = class ApplicationMenu {
...
+      ({ id }) => id === 'Check for Update'
     );
     const checkingForUpdateItem = items.find(
-      ({ label }) => label === 'Checking for Update'
+      // ({ label }) => label === 'Checking for Update'
```
Comment (sadick254) [inline src/main-process/application-menu.js:122]:
```suggestion
```

File: src/main-process/application-menu.js [hunk 9]
```diff
@@ -115,16 +115,20 @@ module.exports = class ApplicationMenu {
...
+      ({ id }) => id === 'Checking for Update'
     );
     const downloadingUpdateItem = items.find(
-      ({ label }) => label === 'Downloading Update'
+      // ({ label }) => label === 'Downloading Update'
```
Comment (sadick254) [inline src/main-process/application-menu.js:126]:
```suggestion
```

File: src/main-process/application-menu.js [hunk 10]
```diff
@@ -115,16 +115,20 @@ module.exports = class ApplicationMenu {
...
+      ({ id }) => id === 'Downloading Update'
     );
     const installUpdateItem = items.find(
-      ({ label }) => label === 'Restart and Install Update'
+      // ({ label }) => label === 'Restart and Install Update'
```
Comment (sadick254) [inline src/main-process/application-menu.js:130]:
```suggestion
```

File: src/menu-manager.coffee [hunk 11]
```diff
@@ -89,6 +89,9 @@ class MenuManager
   #   * `command` An optional {String} command to trigger when the item is
   #     clicked.
   #
+  #   * `id` (internal) The {String} menu id, not used in menu template.
```
Comment (sadick254) [inline src/menu-manager.coffee:92]:
```suggestion
  #   * `id` (internal) A {String} containing the menu item's id.
```

File: src/menu-manager.coffee [hunk 12]
```diff
@@ -89,6 +89,9 @@ class MenuManager
...
+  #     For further information on the `id`, see [Electron MenuItem](https://www.electronjs.org/docs/api/menu-item#class-menuitem)).
+  #
```
Comment (sadick254) [inline src/menu-manager.coffee:94]:
```suggestion
```

**Review by sadick254 — CHANGES_REQUESTED** [review]

File: src/menu-manager.coffee (hunk 5 above)
[the remark this answers is not in the document]
Comment (icecream17) [reply src/menu-manager.coffee:205]: I think you've might forgotten to delete this

[the remark this answers is not in the document]
Comment (icecream17) [reply src/menu-manager.coffee:205]: I think you might've forgotten to delete this

**Comment by icecream17** [comment]
Note: I commented in a resolved suggestion - don't know how to make it show up:
https://github.com/atom/atom/pull/22597#discussion_r663550057

File: src/menu-manager.coffee (hunk 5 above)
[the remark this answers is not in the document]
Comment (juggernautjp) [reply src/menu-manager.coffee:205]: Thank you for your advice. I deleted it on GitHub, but did not delete the line.
And I do not know how to delete it on GitHub. So, I delete it and git push.

**Review by sadick254 — APPROVED** [review]
Thank you for opening this PR and making the changes requested. 🙇

**Pull request merged**